### PR TITLE
K-Means with pytorch 

### PR DESCRIPTION
**Pull request by tgeral68**

### CODE REVIEW

File: geomstats/backend/pytorch.py [hunk 1]
```diff
@@ -285,7 +285,10 @@ def transpose(x, axes=None):
 
 
 def squeeze(x, axis=None):
-    return torch.squeeze(x, dim=axis)
+    if(axis is None):
```
Comment (ninamiolane) [inline geomstats/backend/pytorch.py]: No need for parenthesis here I think

File: geomstats/learning/k_means.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
+import random
+
+import geomstats.backend as gs
+from geomstats.learning._template import TemplateTransformer
+
+
+class KMeans(TemplateTransformer):
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: I'm not sure you should inherit from `TemplateTransformer` here. I think the TemplateTransformer class was written to give you an example of what a child class of TransformerMixin should look like.

In TemplateTransformer, there is 

```class TemplateTransformer(BaseEstimator, TransformerMixin):```

and in the Kmean code, there is:
https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/cluster/_kmeans.py#L649

```class KMeans(TransformerMixin, ClusterMixin, BaseEstimator):```

thus, how about inheriting from TransformerMixin, ClusterMixin, BaseEstimatorm, ie:

```class RiemannianKMeans(TransformerMixin, ClusterMixin, BaseEstimator):```

File: geomstats/learning/k_means.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+    def __init__(self, n_clusters, metric, init="random",
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: single quotes ' instead of "
--> Just because we need to decide on one for code uniformization, and we had chosen ' before.

Comment (ninamiolane) [reply geomstats/learning/k_means.py]: let's put the metric as the first parameter after `self`, as in:
https://github.com/geomstats/geomstats/blob/master/geomstats/learning/pca.py#L100
since it's the only parameter that differs from the scikit-function.

Also, let's set `n_clusters=8` by default, as in scikit-learn?
https://github.com/scikit-learn/scikit-learn/blob/b194674c4/sklearn/cluster/_kmeans.py#L793

File: geomstats/learning/k_means.py [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+        """ K-Means algorithm using Riemannian manifolds
+
+        Parameters
+        ----------
+        n_clusters : Number of clusters (k value of the k-means)
+
+        metric : The geomstats metric associate to the space used
+
+        init : How to init centroids at the beginning of the algorithm.
+               Only radom is implemented yet.
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: NIT: random

File: geomstats/learning/k_means.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+
+    def __init__(self, n_clusters, metric, init="random",
+                 tol=1e-2, verbose=0):
+        """ K-Means algorithm using Riemannian manifolds
+
+        Parameters
+        ----------
+        n_clusters : Number of clusters (k value of the k-means)
+
+        metric : The geomstats metric associate to the space used
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: maybe `Riemannian metric` instead of geomstats metric?

File: geomstats/learning/k_means.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+
+        init : How to init centroids at the beginning of the algorithm.
+               Only radom is implemented yet.
+
+        tol : convergence factor. If the difference of mean distance
+             between two step is lower than tol
+
+        verbose : if verbose > 0, information will be print during learning
+        Returns
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: Cool, thanks for the doc!

Let's just add a line before Returns to separate the inputs and outputs?

File: geomstats/learning/k_means.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        self.n_clusters = n_clusters
+        self.tol = tol
+        self.verbose = verbose
+
+    def fit(self, X, max_iter=100):
+        """Predict for each data point the closest center in terms of metric distance
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: NIT: shape=[n_samples, n_features]

File: geomstats/learning/k_means.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+            # test convergence
+            if(gs.mean(self.metric.dist(old_centroids, self.centroids))
+               < self.tol):
+                print("Convergence Reached after ", index, " iterations")
+                # convergence reached
+                return gs.copy(self.centroids)
+
+    def predict(self, X):
+        # finding closest mean
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: Use the docstring syntax to document the function

File: tests/test_learning_k_means.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
+"""
+Unit tests for landmarks space.
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: Change

File: geomstats/learning/k_means.py [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            Returns the instance itself.
+        """
+        belongs = gs.zeros(X.shape[0])
+        self.centroids = gs.concatenate([gs.expand_dims(X[
+                                         random.randint(0, X.shape[0]-1)], 0)
+                                         for i in range(self.n_clusters)])
+        index = 0
+        while(index < max_iter):
+            index += 1
+            # expectation
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: Comments are difficult to maintain... delete?

File: geomstats/learning/k_means.py [hunk 11]
```diff
@@ -0,0 +1,87 @@
...
+            dists = gs.hstack([self.metric.dist(self.centroids[i], X)
+                              for i in range(self.n_clusters)])
+            belongs = gs.argmin(dists, -1)
+            # maximisation
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: same

File: geomstats/learning/k_means.py [hunk 12]
```diff
@@ -0,0 +1,87 @@
...
+            old_centroids = gs.copy(self.centroids)
+
+            for i in range(self.n_clusters):
+                fold = gs.squeeze(X[belongs == i])
+                if(len(fold) > 0):
+                    self.centroids[i] = self.metric.mean(fold)
+                else:
+                    self.centroids[i] = X[random.randint(0, X.shape[0]-1)]
+
+            # test convergence
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: same

File: geomstats/learning/k_means.py [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+            # expectation
+            dists = gs.hstack([self.metric.dist(self.centroids[i], X)
+                              for i in range(self.n_clusters)])
+            belongs = gs.argmin(dists, -1)
+            # maximisation
+            old_centroids = gs.copy(self.centroids)
+
+            for i in range(self.n_clusters):
+                fold = gs.squeeze(X[belongs == i])
+                if(len(fold) > 0):
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: no parenthesis

File: geomstats/learning/k_means.py [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+    def fit(self, X, max_iter=100):
+        """Predict for each data point the closest center in terms of metric distance
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            Training data, where n_samples is the number of samples
+            and n_features is the number of features.
+
+        max_iter : Maximum number of iteration
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: NIT: iterations

File: geomstats/learning/k_means.py [hunk 15]
```diff
@@ -0,0 +1,87 @@
...
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        belongs = gs.zeros(X.shape[0])
+        self.centroids = gs.concatenate([gs.expand_dims(X[
+                                         random.randint(0, X.shape[0]-1)], 0)
+                                         for i in range(self.n_clusters)])
+        index = 0
+        while(index < max_iter):
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: No parenthesis

File: geomstats/learning/k_means.py [hunk 16]
```diff
@@ -0,0 +1,87 @@
+import random
+
+import geomstats.backend as gs
+from geomstats.learning._template import TemplateTransformer
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: Add line before this line to separate `from` import

File: geomstats/learning/k_means.py [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+
+        max_iter : Maximum number of iteration
+
+        Returns
+        -------
+        self : object
+            Returns the instance itself.
+        """
+        belongs = gs.zeros(X.shape[0])
+        self.centroids = gs.concatenate([gs.expand_dims(X[
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: Can you make this computation in 2 or 3 lines with interpretable variables' names? It would make the code more readable.

File: geomstats/learning/k_means.py [hunk 18]
```diff
@@ -0,0 +1,87 @@
...
+                                         random.randint(0, X.shape[0]-1)], 0)
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: Is there a reason why you use random.randint and not some numpy function?

Comment (ninamiolane) [reply geomstats/learning/k_means.py]: I would also do something like:
```
n_points = X.shape[0]
random_indices = gs.random.randint(n_points, size=self.n_clusters)
self.centroids = X[random_indices]
```
if it runs? It can make it shorter, more readable and factorize X.shape[0] into n_points that you use twice.

File: geomstats/learning/k_means.py [hunk 19]
```diff
@@ -0,0 +1,87 @@
...
+            belongs = gs.argmin(dists, -1)
+            # maximisation
+            old_centroids = gs.copy(self.centroids)
+
+            for i in range(self.n_clusters):
+                fold = gs.squeeze(X[belongs == i])
+                if(len(fold) > 0):
+                    self.centroids[i] = self.metric.mean(fold)
+                else:
+                    self.centroids[i] = X[random.randint(0, X.shape[0]-1)]
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]:
```
random_index = gs.random.randint(n_points)
self.centroids[i] =X[random_index]
```

File: geomstats/learning/k_means.py [hunk 20]
```diff
@@ -0,0 +1,87 @@
...
+
+            # test convergence
+            if(gs.mean(self.metric.dist(old_centroids, self.centroids))
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]:
```
dist_centroids = self.metric.dist(old_centroids, self.centroids)
mean_dist = gs.mean(dist_centroids)

if mean_dist < self.tol:
```
to make it more readable.

File: geomstats/learning/k_means.py [hunk 21]
```diff
@@ -0,0 +1,87 @@
...
+               < self.tol):
+                print("Convergence Reached after ", index, " iterations")
```
Comment (ninamiolane) [inline geomstats/learning/k_means.py]: NIT: R -> r

File: tests/test_learning_k_means.py [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+    def setUp(self):
+        pass
+
+    def test_hyperbolic_space_k_means(self):
+        x = gs.rand(100, 2)
+        manifold = hyperbolic_space.HyperbolicSpace(2)
+        metric = hyperbolic_space.HyperbolicMetric(2)
+        x = manifold.intrinsic_to_extrinsic_coords(x)
+        k_means = KMeans(2, metric, tol=1e-4)
+        # test learning
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: No comment here.

File: tests/test_learning_k_means.py [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+        k_means.fit(x, max_iter=100)
+
+    def test_hypersphere_k_means_fit(self):
+        gs.random.seed(55)
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: Nice!

File: tests/test_learning_k_means.py [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+
+
+class TestLearningKMeans(geomstats.tests.TestCase):
+    _multiprocess_can_split_ = True
+
+    def setUp(self):
+        pass
+
+    def test_hyperbolic_space_k_means(self):
+        x = gs.rand(100, 2)
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: maybe set: `dim = 2` at the beginning, and use it. It will be easier to adapt your code to other dimensions later ;) thanks!

File: tests/test_learning_k_means.py [hunk 25]
```diff
@@ -0,0 +1,65 @@
+"""
+Unit tests for landmarks space.
+"""
+
+import geomstats
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: do you need this import?

File: tests/test_learning_k_means.py [hunk 26]
```diff
@@ -0,0 +1,65 @@
...
+import geomstats.backend as gs
+import geomstats.tests
+
+from geomstats.geometry import hyperbolic_space
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: ```from geomstats.geometry.hyperbolic_space import HyperbolicSpace```
and same for hypersphere.

It's the syntax we have used elsewhere, it's so that the code is shorter when we instantiate a space or a metric

File: tests/test_learning_k_means.py [hunk 27]
```diff
@@ -0,0 +1,65 @@
...
+from geomstats.geometry import hypersphere
+from geomstats.learning.k_means import KMeans
+
+
+class TestLearningKMeans(geomstats.tests.TestCase):
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: Should be:
```
class TestRiemannianKMeansMethods
```
(once you have renamed KMeans to RiemannianKMeans)
to be consistent with the naming in the tests

https://github.com/geomstats/geomstats/blob/master/tests/test_euclidean_space.py#L12

File: tests/test_learning_k_means.py [hunk 28]
```diff
@@ -0,0 +1,65 @@
...
+        pass
+
+    def test_hyperbolic_space_k_means(self):
+        x = gs.rand(100, 2)
+        manifold = hyperbolic_space.HyperbolicSpace(2)
+        metric = hyperbolic_space.HyperbolicMetric(2)
+        x = manifold.intrinsic_to_extrinsic_coords(x)
+        k_means = KMeans(2, metric, tol=1e-4)
+        # test learning
+        k_means.fit(x, max_iter=100)
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: nice, it shows that it runs! Would it be possible to have a test with non-random data so that we can test the output of fit?

File: tests/test_learning_k_means.py [hunk 29]
```diff
@@ -0,0 +1,65 @@
...
+
+    def test_hypersphere_k_means_fit(self):
+        gs.random.seed(55)
+
+        manifold = hypersphere.Hypersphere(2)
+        metric = hypersphere.HypersphereMetric(2)
+
+        x = manifold.random_von_mises_fisher(kappa=100, n_samples=200)
+
+        clustering = KMeans(1, metric)
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: before you called it k_means. maybe decide on a name? also, you're not setting the tol here?

File: tests/test_learning_k_means.py [hunk 30]
```diff
@@ -0,0 +1,65 @@
...
+from geomstats.learning.k_means import KMeans
+
+
+class TestLearningKMeans(geomstats.tests.TestCase):
+    _multiprocess_can_split_ = True
+
+    def setUp(self):
+        pass
+
+    def test_hyperbolic_space_k_means(self):
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: no `_fit` in the name? You have one in the name of the next method.

File: tests/test_learning_k_means.py [hunk 31]
```diff
@@ -0,0 +1,65 @@
...
+
+        x = manifold.random_von_mises_fisher(kappa=100, n_samples=200)
+
+        clustering = KMeans(1, metric)
+        clustering.fit(x)
+        center = clustering.centroids
+        mean = metric.mean(x)
+        result = metric.dist(center, mean)
+        expected = 0.
+        self.assertAllClose(expected, result, atol=1e-2)
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: Nice!

File: tests/test_learning_k_means.py [hunk 32]
```diff
@@ -0,0 +1,65 @@
...
+        metric = hypersphere.HypersphereMetric(2)
+
+        x = manifold.random_von_mises_fisher(kappa=100, n_samples=200)
+
+        k_means = KMeans(5, metric, tol=1e-5)
+        k_means.fit(x, max_iter=100)
+        result = k_means.predict(x)
+
+        centroids = k_means.centroids
+        expected = gs.array([int(metric.closest_neighbor_index(x[i],
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: this is a bit lengthy and hard to read.

you can simplify at least with:
`[ .... x_i .... for x_i in x]`
but probably divide it into 2 lines for readability?

File: tests/test_learning_k_means.py [hunk 33]
```diff
@@ -0,0 +1,65 @@
...
+                                 centroids))
+                             for i in range(len(x))])
+        self.assertAllClose(expected, result)
```
Comment (ninamiolane) [inline tests/test_learning_k_means.py]: Great!

**Comment by ninamiolane** [comment]
Also: you need a description in the PR ;)

**Comment by tgeral68** [comment]
Adding KMeans algorithm for riemannian metrics, small refactor changing vars names, comments and test.

**Comment by tgeral68** [comment]
resolved riemannian k-means and notations
